src/core/abuse-core.js

```diff
@@ -0,0 +1,196 @@
+import _ from "lodash";
+import { GCS_CONFIG } from "../util/gcs";
+const { knex } = require("../util/database").connect();
+
+function _sanitizeText(text) {
+  if (!text) {
+    return text;
+  }
+  return text.replace(/(\n|\r)+/g, " ");
+}
+
+/*
+  Checks if the user has already reported the given feed item,
+  if not, adds a report for it.
+*/
+function reportFeedItem(reportObj) {
+  return knex("feed_item_reports")
+    .select("feed_item_id", "report_creator_id")
+    .where({
+      report_creator_id: reportObj.reportCreatorUuid,
+      feed_item_id: reportObj.feedItemId
+    })
+    .then(check => {
+      if (!_.isEmpty(check)) {
+        throw new Error(
+          "User has already reported this message: " + reportObj.feed_item_id
+        );
+      } else {
+        return knex("users")
+          .where("uuid", reportObj.reportCreatorUuid)
+          .then(users => {
+            if (_.isEmpty(users)) {
+              throw new Error("User not found: " + reportObj.reportCreatorUuid);
+            }
+            const dbRow = {
+              feed_item_id: reportObj.feedItemId,
+              report_creator_id: users[0].uuid,
+              report_description: _sanitizeText(reportObj.reportDescription)
+            };
+            return knex("feed_item_reports")
+              .insert(dbRow)
+              .returning("id")
+              .then(rows => {
+                if (_.isEmpty(rows)) {
+                  throw new Error("Feed item report creation failed: " + dbRow);
+                }
+                return rows.length;
+              })
+              .catch(err => {
+                if (
+                  err.constraint === "feed_item_reports_feed_item_id_foreign"
+                ) {
+                  err.status = 404;
+                  err.message = `No such feed item: ${dbRow.feed_item_id}`;
+                } else if (
+                  err.constraint ===
+                  "feed_item_reports_report_creator_id_foreign"
+                ) {
+                  err.status = 404;
+                  err.message = `No such user: ${dbRow.report_creator_uuid}`;
+                }
+                throw err;
+              });
+          });
+      }
+    });
+}
```

Comment:
There's quite a lot of nested callback chains in here, which is exactly the sort of code Promises are good at getting rid off. Here's an (incomplete) example of how this could be re-formatted to remove the "callback tree" format:

```javascript
return knex("feed_item_reports")
  .select("feed_item_id", "report_creator_id")
  .where({
    report_creator_id: reportObj.reportCreatorUuid,
    feed_item_id: reportObj.feedItemId
  })
  .then(check => {
    if (!_.isEmpty(check)) {
      throw new Error(
        "User has already reported this message: " + reportObj.feed_item_id
      );
    }

    return knex("users")
      .where("uuid", reportObj.reportCreatorUuid);
  })
  .then(users => {
    if (_.isEmpty(users)) {
      throw new Error(`User not found: ${reportObj.reportCreatorUuid}`);
    }

    const dbRow = {
      feed_item_id: reportObj.feedItemId,
      report_creator_id: users[0].uuid,
      report_description: _sanitizeText(reportObj.reportDescription)
    };

    return knex("feed_item_reports")
      .insert(dbRow)
      .returning("id")
  })
  ...
```

This way you'd have to change your `catch` handler to re-throw the non-database errors you've created, but you get (in my opinion) cleaner code and you're utilizing one of the core strengths of Promises, which is promise chaining